### PR TITLE
Yongjian remove admin or owner ability to delete themselves

### PR DESCRIPTION
**Pull request by ypan710**
# Description
Screenshot 2023-09-21 at 1 00 59 PM

## Main changes explained:
- Made admin cannot delete themselves and owner cannot delete themselves nor the admin through the user management functionality.

## How to test:
1. check into current branch
2. do `npm run start:local` to run this PR locally
3. log as admin user
4. go to Dashboard -> Other Links -> User Management -> search for your admin role name -> check to ensure delete button is not visible
5. go to Dashboard -> Other Links -> User Management -> search for your owner role name -> check to ensure delete button is not visible for owner nor admin role

## Screenshots or videos of changes:
For Admin role:

Screenshot 2023-09-21 at 10 55 22 AM

For Owner role:

Screenshot 2023-09-21 at 10 55 58 AM

### CODE REVIEW

**Review by Jiyuanxia — DISMISSED** [review]
Hi @ypan710 , I have reviewed your pull request and tested it. It works as you intended. Well done.

For Admin role:
Screenshot 2023-09-21 171425
For Owner role:
Screenshot 2023-09-21 171628

**Review by Bhagyashree444 — COMMENTED** [review]
Great job, Yongjian! As anticipated, there isn't a delete button available for the Admin account. However, I've noticed that the delete button is still visible for the owner's account.
Screenshot (102)
Screenshot (101)

**Review by wantingxu7 — DISMISSED** [review]
Hi Yongjian, I tested and the functions work perfectly as described. 
For admin:
1331 as admin

For owner:
1331-as owner

**Review by luisarevalo21 — DISMISSED** [review]
Hello @ypan710 ,

I tested your PR and works as intended great work!

Logging in as an admin I cannot see the delete button to delete an admin.
Screenshot 2023-09-21 at 8 01 17 PM

I logged in as an owner and again I cannot see the delete button to delete an owner account. 

Screenshot 2023-09-21 at 8 01 46 PM

**Review by dah00 — DISMISSED** [review]
Hey, PR works expected. Both admin and owner cannot delete their own accounts. Great job!
PR1331_both_admin_and_owner

**Review by Haoj1 — DISMISSED** [review]
@ypan710 Your changes look good to me. Good job!

**Review by DavidC0126 — COMMENTED** [review]
Hi @ypan710, when I tried to run `npm run start`, I failed on 1 task and the ERROR messages are shown in the image. I tried to clear npm cache, delete package.json/node_module and re-install but still failed. Do you have any suggestion on how this error occurred? Thank you!

Screenshot 2023-09-22 at 2 29 59 AM

**Comment by Sssky-18** [comment]
Hi @ypan710, since I don't have an owner account and could not find the "Owner" role to create one, I tested this with my Admin account only. So based on my test, this PR works great. The Admin account cannot delete himself, but still has the ability to delete other volunteer and admin account, and I think it is reasonable. 

PR#1331

**Review by EdwinLauM — DISMISSED** [review]
Hey @ypan710! This PR works as expected! 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/42288987/77756ac4-fcc0-49c5-a8db-bb4254d57378

**Review by harshidad — DISMISSED** [review]
I tested the PR and its working fine. When I logged in with Admin role delete button is not visible and when I logged in with owner role, button is not visible for owner nor any admin role.

Admin Role:
PR#1331_Admin

Owner Role:
PR#1331_OwnerRole

**Review by JeffLi117 — DISMISSED** [review]
I only have an Admin account -- and do not seem to be able to create an Owner account -- so I only tested with Admin. When logging in as Admin, I see your change implemented and am unable to delete myself, AKA works as you intended!

image

**Review by tsunami776 — COMMENTED** [review]
Hi @ypan710 I tested your PR with an admin account. In user management page, my account does not have 'delete' option, which works as described in the PR. But I noticed there are variables been printed in the console. Also, I am not sure if it is only me, or if it is anything related, when I type search words in the top search bar, the page crashes. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/43768723/4c16f36a-82b2-444a-a1eb-114d72b7750b

**Comment by zuhang2000** [comment]
I've tested your PR following your instructions. However I encounter problem when I tried to navigate to the usermanagement page. The following errors reports.
Screenshot 2023-09-22 at 8 27 39 PM

**Comment by one-community** [comment]
Thank you everyone, I'll merge this once the merge conflicts have been resolved.

**Comment by beblicarl** [comment]
This functionality works as intended

### Owner role

image

### Admin role

image

**Comment by one-community** [comment]
Thank you everyone. Fix merge conflicts for this one and I'll merge it!

**Comment by sxiong5** [comment]
Hey @ypan710, I just got a little confused. Shouldn't the owner hold the highest permission? And why the admin can delete the owner?

**Owner**
Screenshot 2023-09-28 at 15 42 49

**Admin**
Screenshot 2023-09-28 at 15 42 20

And btw, you failed the `npm run test` command. Can you check that?

**Comment by Shubhu0** [comment]
Hi @ypan710 Reviewed the PR and it works great! Checked and as seen, the two accounts under my name and the Admin cannot be deleted but the Test Role can be. Good job!

image

**Comment by one-community** [comment]
> Hey @ypan710, I just got a little confused. Shouldn't the owner hold the highest permission? And why the admin can delete the owner?
> 
> **Owner** Screenshot 2023-09-28 at 15 42 49
> 
> **Admin** Screenshot 2023-09-28 at 15 42 20
> 
> And btw, you failed the `npm run test` command. Can you check that?

Good catch! Admin definitely shouldn't be able to delete Owner

**Review by Shiwani99 — COMMENTED** [review]
Hi,
I'm not able to access user management page as it goes blank and throws some error in console.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78906820/39bb4491-e4c1-4713-bcc9-e1cf69654815

**Comment by ypan710** [comment]
> > Hey @ypan710, I just got a little confused. Shouldn't the owner hold the highest permission? And why the admin can delete the owner?
> > **Owner** Screenshot 2023-09-28 at 15 42 49
> > **Admin** Screenshot 2023-09-28 at 15 42 20
> > And btw, you failed the `npm run test` command. Can you check that?
> 
> Good catch! Admin definitely shouldn't be able to delete Owner

@one-community At first I though admin has the highest permission. So admin cannot delete the owner, but can the owner delete the admin? For the failed 'npm run test', this has been happening since I've merged the development branch into my own local branch because of previous permission issues raised by another developer that resulted that I couldn't access the "Other Link" navlink that is required to see the pages you included for the admin and owner above. I can check that.

**Review by ChirayuRai — DISMISSED** [review]
Hey! I just ran your code, and I seem to be getting a lot of errors in the console. BTW, I also tried just adding the useHistory hook into the import statement in UserTableData.jsx, but that didn't fix the issue. Here are the errors I come across when I do try to test this. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/15188751/b2fca813-64d5-4b55-ba8e-b5e68333a9b8

**Review by shubhankarval — DISMISSED** [review]
Hey @ypan710, I'm getting a lot of errors for your PR. I even tried switching to an older branch for the BE to see if that works, but none of the errors are resolving. 
pr

**Comment by ypan710** [comment]
> Hey! I just ran your code, and I seem to be getting a lot of errors in the console. BTW, I also tried just adding the useHistory hook into the import statement in UserTableData.jsx, but that didn't fix the issue. Here are the errors I come across when I do try to test this.
> 
>  Screen.Recording.2023-10-17.at.6.18.11.PM.mov

@ChirayuRai I ran the PR on my side and it seem to be running successfully.
Screenshot 2023-10-21 at 4 05 16 PM

**Comment by ypan710** [comment]
> Hey @ypan710, I'm getting a lot of errors for your PR. I even tried switching to an older branch for the BE to see if that works, but none of the errors are resolving. pr

@shubhankarval I ran the PR on my side and it seem to be running successfully.
Screenshot 2023-10-21 at 4 05 16 PM

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:19]:
```suggestion
```

**Comment by ypan710** [comment]
> Hey @ypan710, I'm getting a lot of errors for your PR. I even tried switching to an older branch for the BE to see if that works, but none of the errors are resolving. pr

@shubhankarval I've commented out the useHistory hook. Please try again. Thanks.

**Comment by ypan710** [comment]
> Hey! I just ran your code, and I seem to be getting a lot of errors in the console. BTW, I also tried just adding the useHistory hook into the import statement in UserTableData.jsx, but that didn't fix the issue. Here are the errors I come across when I do try to test this.
> 
>  Screen.Recording.2023-10-17.at.6.18.11.PM.mov

@ChirayuRai  I've commented out the useHistory hook. Please try again. Thanks.

**Review by Bali2Mohit — COMMENTED** [review]
Hi @ypan710 , I have tested your changes and everything works perfectly.
Can see delete button for my Volunteer account, and not on my Admin account as expected.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94716419/afce8d4d-b17c-453d-bef3-ff3bb9ed9816

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:53]:
```suggestion
      </td>
```
This looks like  typo.

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:112]:
```suggestion
          props.auth.user.userid === props.user._id ? '': 
```
Is this PR specifically to prevent admin and owners from deleting themselves and other roles with permission to delete accounts should be able to delete themselves? Or should everyone (with permission to delete accounts) not be able to delete themselves?

Comment (nathanah) [reply src/components/UserManagement/UserTableData.jsx:112]: This also prevents ANY admin from being deleted which is definitely not the desired behavior

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:37]:
```suggestion
    return props.user.role === 'Owner' && !canAddDeleteEditOwners;
```
`!hasPermission()` will always return false as `hasPermission()` returns a thunk. To get the desired boolean output, you need to call `props.hasPermission()` from `mapDispatchToProps`.

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:24]:
```suggestion
  const canAddDeleteEditOwners = props.hasPermission('addDeleteEditOwners');
```
Remove commented out code and console.logs.
The original usage of `hasPermission()` was correct. The way you changed it makes it so it returns a thunk (which will always equate to true) instead of the desired bool.

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:130]:
```suggestion
              <ResetPasswordButton user={props.user} isSmallButton />
            </span>
```
Fix indentation

**Review by nathanah — DISMISSED** [review]

**Review by sahilpatel18 — DISMISSED** [review]
Although creating an owner role isn't possible for me at the moment, the delete button wasn't visible for the admin account used in testing this PR. Excellent job!

adminSS

**Review by Bali2Mohit — APPROVED** [review]

**Review by nathanah — DISMISSED** [review]
Still has previous problems.

**Review by Alforoan — DISMISSED** [review]
They both worked. Nice!


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/bdf9b241-20d0-41ca-a73c-77d549a0d1e9

**Comment by one-community** [comment]
Thank you all, moving to final review.

File: src/components/UserManagement/UserTableData.jsx
Comment (nathanah) [reply src/components/UserManagement/UserTableData.jsx:36]: Still need this change.

**Review by nathanah — CHANGES_REQUESTED** [review]

**Comment by beblicarl** [comment]
This functionality works as intended for the admin roles but not for the owner roles

Admin role: 

image

Owner role

image

**Review by cvtqx — DISMISSED** [review]
Works as described. When I log in as admin I do not see the delete button next to my admin account name. Same thing for my owner account - I do not see the delete button when logged in as owner. I do see delete buttons next to other accounts but my own. Great job!

Admin:
Admin

Owner:
Owner

**Pull request merged**
